azure-iot-hub-devicesdk/samples/receive_c2d_message.py

```diff
@@ -18,26 +19,26 @@
 device_client = DeviceClient.from_authentication_provider(auth_provider, "mqtt")
 
 
-# The connection state callback allows us to detect when the client is connected and disconnected:
-def connection_state_callback(status):
-    print("connection status: " + status)
+# connect the client.
+device_client.connect()
 
+# enable the device to receive c2d messages
+c2d_message_queue = device_client.get_c2d_message_queue()
 
-def c2d_message_handler(c2d_message):
-    print("the data in the message received was ")
-    print(c2d_message.data)
-    print("custom properties are")
-    print(c2d_message.custom_properties)
 
+# define behavior for receiving a C2D message
+def c2d_listener(message_queue):
+    while True:
+        c2d_message = message_queue.get()  # blocking call
+        print("the data in the message received was ")
+        print(c2d_message.data)
+        print("custom properties are")
+        print(c2d_message.custom_properties)
 
-# Register the connection state callback with the client...
-device_client.on_connection_state = connection_state_callback
 
-# ... and connect the client.
-device_client.connect()
-
-# enable the device to receive c2d messages
-device_client.enable_feature("c2d", c2d_message_handler)
+# Run a listener thread in the background
+listen_thread = threading.Thread(target=c2d_listener, args=(c2d_message_queue,), daemon=True)
```

Comment:
>daemon=True [](start = 81, length = 11)

cool.  I didn't know about the daemon parameter.  This is new for 3.3.  For 2.7 compat, it looks like there's a setDaemon function

Reply:
Oh, good catch, I didn't know it was a different API in 2.7

---
In reply to: [257345281](https://github.com/Azure/azure-iot-sdk-python-preview/pull/49#discussion_r257345281) [](ancestors = 257345281)